src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs

```diff
@@ -18936,5 +18996,1565 @@ interface I1
+                verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).
+            VerifyDiagnostics(
+                // (22,13): warning CS0184: The given expression is never of the provided ('S') type
+                //         if (h2 is S)
+                Diagnostic(ErrorCode.WRN_IsAlwaysFalse, "h2 is S").WithArguments("S").WithLocation(22, 13)
+                );
+
+            // According to
+            // https://github.com/dotnet/runtime/pull/101458#issuecomment-2074169181 and
+            // https://github.com/dotnet/runtime/pull/101458#issuecomment-2075815858
+            // the following is a valid IL
+            verifier.VerifyIL("Helper1<T>.Test1(I1)",
+@"
+{
+  // Code size       22 (0x16)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  isinst     ""T""
+  IL_0006:  brfalse.s  IL_000f
+  IL_0008:  ldc.i4.1
+  IL_0009:  call       ""void System.Console.Write(int)""
+  IL_000e:  ret
+  IL_000f:  ldc.i4.2
+  IL_0010:  call       ""void System.Console.Write(int)""
+  IL_0015:  ret
+}
+");
+
+            verifier.VerifyIL("Helper2.Test2(I1)",
+@"
+{
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldc.i4.4
+  IL_0001:  call       ""void System.Console.Write(int)""
+  IL_0006:  ret
+}
+");
+        }
+
+        [Fact]
+        public void IsOperator_02()
+        {
+            var src1 = @"
+class Helper1<T>
+    where T : allows ref struct
+{
+    public static void Test1(T h1)
+    {
+        if (h1 is I1)
+        {
+        }
+    }
+}
+
+interface I1
+{
+}
+";
+            var comp = CreateCompilation(src1, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+            comp.VerifyDiagnostics(
+                // (7,13): error CS0019: Operator 'is' cannot be applied to operands of type 'T' and 'I1'
+                //         if (h1 is I1)
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "h1 is I1").WithArguments("is", "T", "I1").WithLocation(7, 13)
+                );
+
+            var src2 = @"
+class Helper2
+{
+    public static void Test2(S h2)
+    {
+        if (h2 is I1)
+        {
+            System.Console.Write(3);
+        }
+        else
+        {
+            System.Console.Write(4);
+        }
+    }
+}
+
+ref struct S : I1
+{
+}
+
+interface I1
+{
+}
+
+class Program
+{
+    static void Main()
+    {
+        Helper2.Test2(new S());
+    }
+}
+";
+
+            comp = CreateCompilation(src2, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, options: TestOptions.ReleaseExe);
+            var verifier = CompileAndVerify(
+                comp,
+                expectedOutput: ExecutionConditionUtil.IsMonoOrCoreClr ? "4" : null,
+                verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).
+            VerifyDiagnostics(
+                // (6,13): warning CS0184: The given expression is never of the provided ('I1') type
```

Comment:
This sounds confusing. `S` implements `I1`, so `h2` is of type `I1` in a sense, only the `is` operator cannot be applied to it. Should the diagnostic message be changed to reflect that?

Reply:
> This sounds confusing. `S` implements `I1`, so `h2` is of type `I1` in a sense, only the `is` operator cannot be applied to it. Should the diagnostic message be changed to reflect that?

I would say this is a preexisting condition, which can be demonstrated with a scenario not involving interfaces:
```
warning CS0184: The given expression is never of the provided ('ValueType') type
if (h2 is System.ValueType) 
```